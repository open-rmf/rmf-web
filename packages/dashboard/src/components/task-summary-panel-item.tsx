import React from 'react';
import Debug from 'debug';
import * as RomiCore from '@osrf/romi-js-core-interfaces';
import { Divider, makeStyles, Typography } from '@material-ui/core';
const debug = Debug('OmniPanel:TaskSummary');

export const formatStatus = (status: string) => {
  return status.split('|');
};

export const getStateLabel = (state: number): string => {
  switch (state) {
    case RomiCore.TaskSummary.STATE_QUEUED:
      return 'QUEUED';
    case RomiCore.TaskSummary.STATE_ACTIVE:
      return 'ACTIVE';
    case RomiCore.TaskSummary.STATE_COMPLETED:
      return 'COMPLETED';
    case RomiCore.TaskSummary.STATE_FAILED:
      return 'FAILED';
    default:
      return 'UNKNOWN';
  }
};
interface TaskSummaryPanelItemProps {
  task: RomiCore.TaskSummary;
}

export const TaskSummaryPanelItem = React.memo(
  React.forwardRef(function (
    props: TaskSummaryPanelItemProps,
    ref: React.Ref<HTMLElement>,
  ): React.ReactElement {
    debug('render');
    const { task } = props;
    const classes = useStyles();

    const statusDetails = formatStatus(task.status);
    const stateLabel = getStateLabel(task.state);

    return (
      <>
        <div className={classes.accordionDetailLine}>
          <Typography variant="body1">TaskId: </Typography>
          <Typography variant="body1" noWrap>
            {task.task_id}
          </Typography>
        </div>
        <Divider />

        <div className={classes.accordionDetailLine}>
          <Typography variant="body1">State:</Typography>
          <Typography variant="body1">{stateLabel}</Typography>
        </div>
        <Divider />

        <div className={classes.accordionDetailLine}>
          <Typography variant="body1">Status:</Typography>
          <div>
            {statusDetails.map((detail: string) => {
              return (
                <Typography variant="body1" key={detail}>
                  {' '}
                  {detail}
                </Typography>
              );
            })}
          </div>
        </div>
        <Divider />

        <div className={classes.accordionDetailLine}>
          <Typography variant="body1">Submission Time:</Typography>
          <Typography variant="body1">{task.submission_time.sec}</Typography>
        </div>
        <Divider />

        <div className={classes.accordionDetailLine}>
          <Typography variant="body1">Start time:</Typography>
          <Typography variant="body1">{task.start_time.sec}</Typography>
        </div>
        <Divider />

        <div className={classes.accordionDetailLine}>
          <Typography variant="body1">End Time:</Typography>
          <Typography variant="body1">{task.end_time.sec}</Typography>
        </div>
      </>
    );
  }),
);

const useStyles = makeStyles((theme) => ({
  accordionDetailLine: {
    display: 'flex',
    justifyContent: 'space-between',
    padding: theme.spacing(0.5),
  },
}));
