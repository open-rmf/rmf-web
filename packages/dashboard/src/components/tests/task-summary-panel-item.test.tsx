import { createMount } from '@material-ui/core/test-utils';
import * as RomiCore from '@osrf/romi-js-core-interfaces';
import { shallow } from 'enzyme';
import React from 'react';
import fakeTaskSummary from '../../mock/data/task-summary';
import { TaskSummaryPanelItem } from '../task-summary-panel-item';

const mount = createMount();

it('Matches snapshot', () => {
  const task = Object.values(fakeTaskSummary())[0];
  const root = shallow(<TaskSummaryPanelItem task={task} />);
  expect(root).toMatchSnapshot();
  root.unmount();
});

describe('Components gets the correct label on specifics states', () => {
  let task: RomiCore.TaskSummary;
  beforeEach(() => {
    task = Object.values(fakeTaskSummary())[0];
  });

  test('Shows ACTIVE label', () => {
    task.state = RomiCore.TaskSummary.STATE_ACTIVE;
    const root = mount(<TaskSummaryPanelItem task={task} />);
    expect(root.html()).toContain('ACTIVE');
    root.unmount();
  });

  test('Shows QUEUE label', () => {
    task.state = RomiCore.TaskSummary.STATE_QUEUED;
    const root = mount(<TaskSummaryPanelItem task={task} />);
    expect(root.html()).toContain('QUEUE');
    root.unmount();
  });

  test('Shows COMPLETED label', () => {
    task.state = RomiCore.TaskSummary.STATE_COMPLETED;
    const root = mount(<TaskSummaryPanelItem task={task} />);
    expect(root.html()).toContain('COMPLETED');
    root.unmount();
  });

  test('Shows FAILED label', () => {
    task.state = RomiCore.TaskSummary.STATE_FAILED;
    const root = mount(<TaskSummaryPanelItem task={task} />);
    expect(root.html()).toContain('FAILED');
    root.unmount();
  });
});
