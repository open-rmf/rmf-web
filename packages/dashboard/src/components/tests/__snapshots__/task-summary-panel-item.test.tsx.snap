// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<Fragment>
  <div
    className="makeStyles-accordionDetailLine-1"
  >
    <WithStyles(ForwardRef(Typography))
      variant="body1"
    >
      TaskId: 
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      noWrap={true}
      variant="body1"
    >
      af8faee9-84ca-41ea-8bb6-8493cc9f824c
    </WithStyles(ForwardRef(Typography))>
  </div>
  <WithStyles(ForwardRef(Divider)) />
  <div
    className="makeStyles-accordionDetailLine-1"
  >
    <WithStyles(ForwardRef(Typography))
      variant="body1"
    >
      State:
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body1"
    >
      ACTIVE
    </WithStyles(ForwardRef(Typography))>
  </div>
  <WithStyles(ForwardRef(Divider)) />
  <div
    className="makeStyles-accordionDetailLine-1"
  >
    <WithStyles(ForwardRef(Typography))
      variant="body1"
    >
      Status:
    </WithStyles(ForwardRef(Typography))>
    <div>
      <WithStyles(ForwardRef(Typography))
        key="Moving [tinyRobot/tinyRobot1]: ( 9.81228 -6.98942 -3.12904) -> ( 6.26403 -3.51569  1.16864) "
        variant="body1"
      >
         
        Moving [tinyRobot/tinyRobot1]: ( 9.81228 -6.98942 -3.12904) -&gt; ( 6.26403 -3.51569  1.16864) 
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        key=" Remaining phases: 1 "
        variant="body1"
      >
         
         Remaining phases: 1 
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        key=" Remaining phases: 6"
        variant="body1"
      >
         
         Remaining phases: 6
      </WithStyles(ForwardRef(Typography))>
    </div>
  </div>
  <WithStyles(ForwardRef(Divider)) />
  <div
    className="makeStyles-accordionDetailLine-1"
  >
    <WithStyles(ForwardRef(Typography))
      variant="body1"
    >
      Submission Time:
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body1"
    >
      0
    </WithStyles(ForwardRef(Typography))>
  </div>
  <WithStyles(ForwardRef(Divider)) />
  <div
    className="makeStyles-accordionDetailLine-1"
  >
    <WithStyles(ForwardRef(Typography))
      variant="body1"
    >
      Start time:
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body1"
    >
      0
    </WithStyles(ForwardRef(Typography))>
  </div>
  <WithStyles(ForwardRef(Divider)) />
  <div
    className="makeStyles-accordionDetailLine-1"
  >
    <WithStyles(ForwardRef(Typography))
      variant="body1"
    >
      End Time:
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body1"
    >
      0
    </WithStyles(ForwardRef(Typography))>
  </div>
</Fragment>
`;
